src/engine/enginemaster.h

```diff
@@ -126,68 +128,104 @@ class EngineMaster : public QObject, public AudioSource {
+        bool m_fadeout;
     };
 
     class GainCalculator {
       public:
         virtual double getGain(ChannelInfo* pChannelInfo) const = 0;
     };
-    class ConstantGainCalculator : public GainCalculator {
+    class PflGainCalculator : public GainCalculator {
       public:
         inline double getGain(ChannelInfo* pChannelInfo) const {
-            return pChannelInfo->m_pChannel->isTalkover() ? m_dTalkoverGain : m_dGain;
+            Q_UNUSED(pChannelInfo);
+            return m_dGain;
         }
         inline void setGain(double dGain) {
             m_dGain = dGain;
         }
-        inline void setTalkoverGain(double dGain) {
-            m_dTalkoverGain = dGain;
-        }
       private:
         double m_dGain;
-        double m_dTalkoverGain;
+    };
+    class TalkoverGainCalculator : public GainCalculator {
+      public:
+        inline double getGain(ChannelInfo* pChannelInfo) const {
+            Q_UNUSED(pChannelInfo);
+            return 1.0;
+        }
     };
     class OrientationVolumeGainCalculator : public GainCalculator {
       public:
         OrientationVolumeGainCalculator()
-                : m_dVolume(1.0), m_dLeftGain(1.0), m_dCenterGain(1.0), m_dRightGain(1.0),
-                  m_dTalkoverGain(1.0) { }
+                : m_dVolume(1.0),
+                  m_dLeftGain(1.0),
+                  m_dCenterGain(1.0),
+                  m_dRightGain(1.0) {
+        }
 
         inline double getGain(ChannelInfo* pChannelInfo) const {
-            if (pChannelInfo->m_pMuteControl->get() > 0.0) {
-                return 0.0;
-            }
-            if (pChannelInfo->m_pChannel->isTalkover()) {
-                return m_dTalkoverGain;
-            }
             const double channelVolume = pChannelInfo->m_pVolumeControl->get();
             const double orientationGain = EngineMaster::gainForOrientation(
-                pChannelInfo->m_pChannel->getOrientation(),
-                m_dLeftGain, m_dCenterGain, m_dRightGain);
+                    pChannelInfo->m_pChannel->getOrientation(),
+                    m_dLeftGain, m_dCenterGain, m_dRightGain);
             return m_dVolume * channelVolume * orientationGain;
         }
 
-        inline void setGains(double dVolume, double leftGain, double centerGain, double rightGain,
-                             double talkoverGain) {
+        inline void setGains(double dVolume, double leftGain,
+                double centerGain, double rightGain) {
             m_dVolume = dVolume;
             m_dLeftGain = leftGain;
             m_dCenterGain = centerGain;
             m_dRightGain = rightGain;
-            m_dTalkoverGain = talkoverGain;
         }
 
       private:
-        double m_dVolume, m_dLeftGain, m_dCenterGain, m_dRightGain, m_dTalkoverGain;
+        double m_dVolume;
+        double m_dLeftGain;
+        double m_dCenterGain;
+        double m_dRightGain;
+    };
+    template<typename T, unsigned int CAPACITY>
+    class FastVector {
```

Comment:
> The performance gain might be minor, but there is one. 

Software development is a balancing act of different concerns. I'm coming at this from a maintenance perspective.

There are all sorts of unmaintainable hacks we could do to the engine that would be a performance gain at the expense of other goals. 

> We come from a highly optimized solution using 32 bit flags + a plain array with unrolled sampleutil_generated copy functions, that squeezes every unnecessary instruction. In one of my first version I had a plain array + a size integer. It was just natural to combine them into a class.
> If we switch to a QVarLengthArray we will have up to 5 * 32 = 160 limit checks when appending and 10 * 32 = 320 checks when accessing the vector on every callback. This does not fit to the rest of the code.

Yes, I wrote that flag code and the auto-generation functions. 

Performance is not a focus in a debug build (Mixxx and every library is built with O0) so let's drop the access question since they are identical in a release build.

To help measure the cost involved I wrote a benchmark. FastVector::append's raw performance in a tight loop on a 32-element pointer container is roughly 25% faster than QVarLengthArray as measured over 1 million iterations with PerformanceTimer. 1.87 nanoseconds per append versus 1.39 nanoseconds per append on my machine.

So, 160 appends would mean a cost of about 76ns per callback on my machine. That's 0.0076% of the callback budget at 1ms latency. This is quite small, especially compared to the ChannelMixer methods that immediately follow this code.

I benchmarked access using ```operator[]``` for QVarLengthArray and FastVector and the were identical (.074ns per access over 1e6 iterations of a full vector) with:

```c++
inline void noop() { }
#define Q_ASSERT(x) noop()
```

> More code to maintain -> more bugs to fix.
> I think the FastVector is already one of the best reviewed code.

Except if anyone tries to use it as a general vector they're going to run into bugs if they don't know how it works. "FastVector" hints that it's just like a vector (which conventionally refers to an auto-resizing array).

> Memory corruption crashes instead of clean assertion failures in debug builds.
> Do not get the point, here. We have limit checks outside the Vector.

In this usage of the class, yes. But what about the next developer who says "I obviously want the fast kind of vector so I'll use this nice FastVector class!"

> Why that? I am against introducing a dynamic channel count. This will come with CPU costs, for all users, where only exotic use-cases have a benefit. I am sure we can find a suitable fixed maximum count.

Ah, I was mistaken. ChannelMixer already has the fallback >N channel case. This comes at no CPU cost to non-exotic uses -- it's a fall-back. 

Correctness matters more than speed and the bitvector limitation I added in v1.7 made sense at the time when we had 2 hard-coded decks and 4 hard-coded samplers but it no longer makes sense today when users can viably create a 128 sampler skin without any code changes.

We've wasted too much time on this discussion -- let's go ahead and merge. Once I figure out the best way to disable Q_ASSERT in our builds I'll remove these.